src/net/ossl.cc

```diff
@@ -590,249 +640,446 @@ class session : public enable_shared_from_this<session>, public session_impl {
-                return make_exception_future<buf_type>(_error);
+            } else {
+                buf.trim(bytes_read);
+                return make_ready_future<buf_type>(std::move(buf));
             }
-            buf.trim(bytes_read);
-            return make_ready_future<buf_type>(std::move(buf));
         });
     }
 
+    // Called by user applications to pull data off of the TLS session
     future<buf_type> get() override {
         if (_error) {
-            return make_exception_future<temporary_buffer<char>>(_error);
+            return make_exception_future<buf_type>(_error);
         }
         if (_shutdown || eof()) {
-            return make_ready_future<temporary_buffer<char>>(buf_type());
+            return make_ready_future<buf_type>(buf_type());
         }
         if (!connected()) {
             return handshake().then(std::bind(&session::get, this));
         }
-        return with_semaphore(_in_sem, 1, std::bind(&session::do_get, this)).then([](temporary_buffer<char> buf) {
-            // TODO(rob) - maybe re-handshake?
-            return make_ready_future<temporary_buffer<char>>(std::move(buf));
-        });
+        return with_semaphore(_in_sem, 1, std::bind(&session::do_get, this))
+          .then([this](buf_type buf) {
+              if (buf.empty() && !eof()) {
+                  return handshake().then(std::bind(&session::get, this));
+              }
+              return make_ready_future<buf_type>(std::move(buf));
+          });
     }
 
+    // Performs shutdown
     future<> do_shutdown() {
-        if(_error || !connected() || eof()) {
+        if (_error || !connected()) {
             return make_ready_future();
         }
+
         auto res = SSL_shutdown(_ssl.get());
-        if (res == 1){
-            // Shutdown has completed successfully
-            return make_ready_future<>();
+        if (res == 1) {
+            return make_ready_future();
         } else if (res == 0) {
-            // Shutdown process is ongoing and has not yet completed, peer has not yet replied
-            // 0 does not indicate error, calling SSL_get_error is undefined
-            return yield().then([this]{
-                return do_shutdown();
-            });
-        }
-        // Shutdown was not successful, calling SSL_get_error will indicate why
-        auto err = SSL_get_error(_ssl.get(), res);
-        if (err == SSL_ERROR_WANT_READ) {
-            auto f = make_ready_future();
-            if (_type == session_type::CLIENT) {
-                // Clients will be sending the close_notify message, and expecting SSL_ERROR_ZERO_RETURN
-                // from the server, logic in wait_for_input will detect this and set _eof to true
-                f = pull_encrypted_and_send();
-            }
-            return f.then([this]{
-                return wait_for_input().then([this] {
-                    return do_shutdown();
+            return yield().then([this] { return do_shutdown(); });
```

Comment:
> admittedly, I have no clue how long the `SSL_shutdown` routine takes to complete, but it seems like the recursion here could go fairly deep. would it make sense to drive the polling checks iteratively?

The yield returns control back to the scheduler, sending the continuation to the back of the queue and preventing deep recursion, so it's kind of a trampoline, here. I think the scheduler also keeps track of the number of direct invocations of ready futures and will force a yield, but it's better to be explicit here since we know it's not ready.

Reply:
ah, yeah of course it works that way.